### DISCUSSION TITLE
Delaying the deletion of draft orders

### DISCUSSION THREAD

**Discussion by abishekrsrikaanth** [woocommerce/woocommerce#54097] [WooCommerce, Blocks, and Site Editing]
The draft orders get deleted out of the box every 24 hours. We create a draft order for the abandoned carts so when the customer comes back, they can continue to checkout using the draft order. This is causing an issue with our abandoned carts and would like to delay the deletion of the draft orders. 

Here is a snippet that I tried based on a suggestion from a developer on the Slack Channel, but that did not help.
```php
<?php

// 1. Deactivate the default WooCommerce cleanup draft orders cron
add_action( 'init', 'deactivate_woocommerce_cleanup_draft_orders' );
function deactivate_woocommerce_cleanup_draft_orders() {

    if ( function_exists( 'as_next_scheduled_action' ) ) {
        $timestamp = as_next_scheduled_action( 'woocommerce_cleanup_draft_orders' );
        
        if ( $timestamp ) {
            as_unschedule_action( 'woocommerce_cleanup_draft_orders' );
        }
    }
}

// 2. Schedule a custom cron job for cleaning draft orders every 3 days
add_action( 'init', 'schedule_custom_cleanup_draft_orders' );
function schedule_custom_cleanup_draft_orders() {
    if ( function_exists( 'as_next_scheduled_action' ) ) {
        $timestamp = as_next_scheduled_action( 'woocommerce_cleanup_draft_orders' );
        if ( ! $timestamp ) {
            as_schedule_recurring_action( strtotime( 'midnight tonight' ), 3 * DAY_IN_SECONDS, 'woocommerce_cleanup_draft_orders' );
        }
    }
}
```

I added this as a code snippet and this did not help. Any advice on what other options are there to add a delay for about 3 days to the draft orders..

**Comment by nielslange** [comment]
Hello @abishekrsrikaanth 👋

Thanks for reaching out. Technically, the code snippet should work. On my end, I see the correct schedule:

<img width="1109" alt="Screenshot 2025-01-07 at 16 31 06" src="https://github.com/user-attachments/assets/43befca9-7d03-40ba-8c21-7c9c3a2b854c" />

I've added your code snippet using the plugin [Code Snippets](https://wordpress.org/plugins/code-snippets/) and I've looked up the scheduled actions using [WP Crontrol](https://id.wordpress.org/plugins/wp-crontrol/).

I suggest running [a plugin and theme conflict test](https://woocommerce.com/document/how-to-test-for-conflicts/) to check if the issue you are facing is caused by your theme or one of your activated plugins. I hope this helps to solve this issue.

**Comment by coreymckrill** [comment]
@abishekrsrikaanth a couple other thoughts here in addition to the above.

* [The code](https://github.com/woocommerce/woocommerce/blob/d1498d45f2804eede99538d13717c4e2cf602308/plugins/woocommerce/src/Blocks/Domain/Services/DraftOrders.php#L63-L67) that schedules the action appears to run in `admin_init`, which is after `init`, which might be why your code isn't working. You could try switching your action hook to `admin_init` with a high priority number, to make sure it runs after the core code.

* The snippet you're using could actually end up leaving some drafts sitting indefinitely. Instead, you could try filtering [the query args](https://github.com/woocommerce/woocommerce/blob/d1498d45f2804eede99538d13717c4e2cf602308/plugins/woocommerce/src/Blocks/Domain/Services/DraftOrders.php#L162-L169) for the query that retrieves the draft orders that are going to get deleted, so that they take longer to expire. Something like this (not tested, use at your own risk):

```php
add_filter(
    'woocommerce_order_query_args',
    function( $args ) {
        if ( ! doing_action( 'woocommerce_cleanup_draft_orders' ) || ! isset( $args['status'] ) || 'wc-checkout-draft' !== $args['status'] || ! isset( $args['date_modified'] ) ) {
            return $args;
        }

        $args['date_modified'] = '<=' . strtotime( '-3 DAY' );
        return $args;
    }
);
```

**Reply by abishekrsrikaanth** [reply]
@coreymckrill  I am using Code Snippets app to run the above code. How do I go about doing this? Would have been great if there was some kind of configuration as part of WooCommerce's General Settings to indicate how many days before which draft orders should be deleted instead of just deleting after 24 hours... 

Having said that, what should I set the priority based on my current setup or how do I go about fixing based on what you mentioned above. Can I still use the `Code Snippets` app to fix this?

![CleanShot 2025-01-12 at 11 22 39@2x](https://github.com/user-attachments/assets/d499dec1-daba-4595-91a5-b0a2cecb6c8e)

**Reply by coreymckrill** [reply]
@abishekrsrikaanth you could replace the snippet you have with the one a mentioned above, but again, I haven't tested it so I would not recommend using it on a live site. Try it on a staging site or other test site first.
